internal/creds.go

```diff
@@ -103,3 +115,15 @@ func QuotaProjectFromCreds(cred *google.Credentials) string {
 	}
 	return v.QuotaProject
 }
+
+func impersonateCredentials(ctx context.Context, creds *google.Credentials, ds *DialSettings) (*google.Credentials, error) {
+	if len(ds.ImpersonationConfig.Scopes) == 0 {
+		ds.ImpersonationConfig.Scopes = ds.Scopes
+	}
+	ts, err := impersonate.TokenSource(ctx, creds.TokenSource, ds.ImpersonationConfig)
+	if err != nil {
+		return nil, err
+	}
+	creds.TokenSource = ts
```

Comment:
wonder if creds.JSON needs to be niled out here.

Reply:
I think you should probably return a brand new *google.Credentials

```
return &google.Credentials{ TokenSource: ts }
```

Reply:
You are probably right about the JSON, but maybe we still want the ProjectId field to be copied through?